### PR TITLE
Pinning of the AUR's TLS public key

### PR DESCRIPTION
**Pull request by clawoflight**
This is WIP towards implementing #185.

- [x] Patch aursync
- [ ] Patch aursearch
- [ ] Patch aurgrep
- [ ] Document

### CODE REVIEW

File: bin/aursync [hunk 1]
```diff
@@ -53,12 +64,12 @@ download() {
     declare gzdir=$1 startdir=$PWD
     cd_safe "$gzdir"
 
-    if type -P aria2c >/dev/null 2>&1; then
+    if ((pubkeypin==0)) && type -P aria2c >/dev/null 2>&1; then
         aria2c --stderr=true -i -
     elif type -P parallel >/dev/null 2>&1; then
-        parallel --will-cite -X -j +5 --nice 10 --halt soon,fail=1 "curl -fLsS --remote-name-all {}"
+        parallel --will-cite -X -j +5 --nice 10 --halt soon,fail=1 "curl ${curl_args[*]} {}"
```
Comment (clawoflight) [inline bin/aursync]: I am unsure whether this is the correct expansion to use.
Shellcheck gave me an error when I used `[@]` here.

File: bin/aursync [hunk 2]
```diff
@@ -33,6 +37,13 @@ clone() {
                 warning "remote $remote does not match $repo"
             fi
 
+            # Update the config to get the newest pubkey. Is this smart or not efficient enough?
```
Comment (clawoflight) [inline bin/aursync]: I thought about this while writing the other stuff.

`time` tells me the performance impact is negligible, but that is on a fast machine and you or others might disagree.

File: bin/aursync (hunk 1 above)
Comment (AladW) [inline bin/aursync]: [*] is the right way; e.g. try `foo=(a b c); printf '%s\n' "foo: ${foo[@]}" && printf '%s\n' "${foo[*]}"`. You can find a more detailed description in the Shellcheck wiki

File: bin/aursync (hunk 2 above)
Comment (AladW) [inline bin/aursync]: It's something like 0.003s for me... should be negligable compared to the other stuff.

File: bin/aursync (hunk 1 above)
Comment (clawoflight) [inline bin/aursync]: Thanks, that example cleared it up :)

File: bin/aursync (hunk 2 above)
Comment (clawoflight) [inline bin/aursync]: Cool. That removes the need for a way to manually update the hash.

**Comment by clawoflight** [comment]
Should I start documenting this, or are you waiting for something? :)

**Comment by AladW** [comment]
It's more that I'm hesitant knowing there's no upstream support for actual HPKP, even when client-side pinning works fine. If you could do `--cert-status` in a separate commit, I'd merge that one directly, though.

**Comment by clawoflight** [comment]
I could do that, but it's not really worth a PR. You can add that to the two calls two `curl` if you want :)

Do you have a specific reason to hesitate? :)
I actually prefer client-side pinning because it's an out-of-band trust model, whereas everything is sent along with HPKP.

**Comment by clawoflight** [comment]
Cool, github kept the comments after a rebase and force-push :)

**Comment by AladW** [comment]
My concern for client-side pinning when HPKP is not supported by the site is, how do you make sure you don't have a fake, MITM certificate at the time you set the public key in your configuration? (I suppose you have a similar issue with HPKP, but only on the very first connection to the website)

**Comment by clawoflight** [comment]
> how do you make sure you don't have a fake MITM certificate at the time you set the public key in your configuration? 

By thinking and being careful. For example, you could
* check several times during the week, and/or
* check from within different networks, e.g. uni, and/or
* use different DNS servers, and/or
* also check through a VPN, and/or
* check after SSHing somewhere else.

Maybe the admin of the machine the AUR runs on would even GPG-sign a hash generated directly from the file and/or post it to aur-general if you ask nicely :)

I don't find this inherently less secure than HPKP, but rather a trade-off. HPKP is nice because it requires no user-interaction, but a MITM on first connection negates it completely.
Whereas this makes out-of-channel checks much easier, and even allows you to *secure a new machine before ever connecting to the internet*, simply by copying a trusted hash over from another one.
Or, for that matter, pass a hash on to close friends, or compare against theirs.

**Comment by clawoflight** [comment]
For reference, I have this in my `~/.zshrc`:

```zsh
function generate-tls-pubkey-hash() {
	openssl s_client -servername $1 -connect $1:443 |
	 openssl x509 -pubkey -noout |
	 tee /dev/stderr |
	 openssl rsa -pubin -outform der |
	 openssl dgst -sha256 -binary |
	 openssl enc -base64
}
```

The resulting hash can be used in curl by prepending `sha256//`.

**Comment by clawoflight** [comment]
Did you make up your mind on this? :)

**Comment by AladW** [comment]
Sorry, I don't have much time to look at aurutils atm

**Comment by clawoflight** [comment]
Okay, thanks for the reply, just wanted to get a lifesign :)

**Comment by AladW** [comment]
See https://github.com/AladW/aurutils/issues/185#issuecomment-326458202. If you want to implement this in a separate repo, let me know and I'll link it from the readme.

**Pull request closed without merging**